components/brave_ads/browser/ads_service_impl.cc

```diff
@@ -0,0 +1,933 @@
+
+  file_task_runner_->PostTask(FROM_HERE,
+      base::BindOnce(&EnsureBaseDirectoryExists, base_path_));
+  profile_pref_change_registrar_.Init(profile_->GetPrefs());
+  profile_pref_change_registrar_.Add(
+      prefs::kBraveAdsEnabled,
+      base::Bind(&AdsServiceImpl::OnPrefsChanged,
+                 base::Unretained(this)));
+  profile_pref_change_registrar_.Add(
+      prefs::kBraveAdsIdleThreshold,
+      base::Bind(&AdsServiceImpl::OnPrefsChanged,
+                 base::Unretained(this)));
+
+  auto* display_service_impl =
+      static_cast<NotificationDisplayServiceImpl*>(display_service_);
+
+  display_service_impl->AddNotificationHandler(
+      NotificationHandler::Type::BRAVE_ADS,
+      std::make_unique<AdsNotificationHandler>(this));
+
+  if (is_enabled())
+    Start();
+  else
+    Stop();
+}
+
+AdsServiceImpl::~AdsServiceImpl() {
+  file_task_runner_->DeleteSoon(FROM_HERE, bundle_state_backend_.release());
+}
+
+void AdsServiceImpl::OnInitialize() {
+  enabled_ = true;
+  ResetTimer();
+}
+
+void AdsServiceImpl::OnCreate() {
+  if (!bat_ads_.is_bound()) {
+    // TODO(bridiver) - reconnect?
+    return;
+  }
+
+  bat_ads_->Initialize(
+      base::BindOnce(&AdsServiceImpl::OnInitialize, AsWeakPtr()));
+}
+
+void ConnectionClosed() {
+  // TODO(bridiver)
+}
+
+void AdsServiceImpl::Start() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  DCHECK(is_enabled());
+  DCHECK(!bat_ads_.is_bound());
+
+  bat_ads::mojom::BatAdsClientAssociatedPtrInfo client_ptr_info;
+  bat_ads_client_binding_.Bind(mojo::MakeRequest(&client_ptr_info));
+
+  content::ServiceManagerConnection* connection =
+      content::ServiceManagerConnection::GetForProcess();
+
+  if (!connection)
+    return;
+
+  connection->GetConnector()->BindInterface(
+      bat_ads::mojom::kServiceName, &bat_ads_service_);
+
+  bat_ads_service_.set_connection_error_handler(
+      base::Bind(&ConnectionClosed));
+
+  bool is_production = false;
+#if defined(OFFICIAL_BUILD)
+  is_production = true;
+#endif
+  bool is_debug = true;
+#if defined(NDEBUG)
+  is_debug = false;
+#endif
+  bool is_testing = false;
+
+  const base::CommandLine& command_line =
+      *base::CommandLine::ForCurrentProcess();
+
+  if (command_line.HasSwitch(switches::kProduction)) {
+    is_production = true;
+  }
+  if (command_line.HasSwitch(switches::kDebug)) {
+    is_debug = true;
+  }
+  if (command_line.HasSwitch(switches::kTesting)) {
+    is_testing = true;
+  }
+
+  bat_ads_service_->SetProduction(is_production, base::BindOnce(&Noop));
+  bat_ads_service_->SetDebug(is_debug, base::BindOnce(&Noop));
+  bat_ads_service_->SetTesting(is_testing, base::BindOnce(&Noop));
+
+  bat_ads_service_->Create(std::move(client_ptr_info), MakeRequest(&bat_ads_),
+      base::BindOnce(&AdsServiceImpl::OnCreate, AsWeakPtr()));
+}
+
+void AdsServiceImpl::Stop() {
+  enabled_ = false;
+  Shutdown();
+}
+
+void AdsServiceImpl::ResetTimer() {
+  idle_poll_timer_.Stop();
+  idle_poll_timer_.Start(FROM_HERE,
```

Comment:
sorry, this was meant for `SetTimer` not `ResetTimer`

Reply:
i remember pointing this out in a previous PR so i think there is already a method defined for doing something every T interval with a randomized delay each time

Reply:
to clarify, this only applies to cases where the thing-which-is-recurring involves a network request. ex: fetching ad catalog, reporting confirmations. it does not apply to local-only operations like picking an ad to show out of the local catalog or running a local classifier.

Reply:
> i remember pointing this out in a previous PR so i think there is already a method defined for doing something every T interval with a randomized delay each time

Could you please point me to the method (or if you know the the randomized delay range) and I will implement the same logic on the native-ads for fetching the catalog

Reply:
@tmancey i was thinking of https://github.com/brave/brave-core/commit/0649d184e33c070625c2a537d70a95ec662100ef, which is not reusable and also is not really correct (see my comment on the commit)

basically you want the equivalent of this (in JS):

```
function doThingWithRandomDelay(interval, maxRandomOffset, callback) {
  return setInterval(setTimeout(callback, maxRandomOffset), interval)
}
```

where `maxRandomOffset` is at least interval divided by expected-number-of-requests-per-interval across all clients.

for instance, say there are 100,000 daily active users which make one of these requests per 24 hours. 24 hours divided by 100,000 is 0.864s. then 2s is a reasonable value for maxRandomOffset.

to see why this is necessary, consider the case where there are only two users of Brave Ads, Alice and Bob:
* Alice makes her first ads catalog request at 6:00:00 and repeats it exactly every 24hours
* Bob makes his first ads catalog request at 18:00:00 and repeats it exactly every 24hours
* Brave can easily distinguish the two users based on when the request is occurring

granted this is not a big deal for ads catalog since that request shouldn't include any sensitive info, but it is important for ads confirmation (in which we want requests from the same user to be as unlinkable as possible)

Reply:
Ads confirmation is not added yet and will start to be added next week. This does make sense and based upon above I now need to find out the expected number of requests per interval across all clients. Is this something we can address once ads confirmation is added?

Reply:
@tmancey sgtm; please open a follow up issue if one is needed

Reply:
@diracdeltas Right now we check for and download a new catalog every x seconds, determined by a PING value returned from the server. If for any reason a check fails we then on desktop retry after 1 minute (or mobile 2 minutes) and if that fails we we multiply the retry by 2 until we are successful (or the app is restarted where we revert to the PING value). Could you please let me know the random -/+ ve to be to cater for the above, thanks

Reply:
@tmancey what is the range of x?

Reply:
x can be any number and is passed back in the JSON of the catalog, i.e. 7200000 which in seconds equates to 7200 (2 hours) right now. Thanks

Reply:
@tmancey `x/10` seems reasonable to me

Reply:
👍 making the change now, and will update once complete

Reply:
@diracdeltas fix pushed and code is as follows:

```void AdsServe::UpdateNextCatalogCheck() {
  next_retry_start_timer_in_ = 0;

  auto ping = bundle_->GetCatalogPing();
  auto rand_delay = helper::Math::Random(ping / 10);

  // Add randomized delay so that the Ad server can't correlate users by timing
  next_catalog_check_ = ping + rand_delay;

  ads_->StartCollectingActivity(next_catalog_check_);
}
```